File: packages/types/src/events/system-events.ts
/**
 * System Event Type Definition
 */

import type { ID } from "../common.js";
import type { BaseEvent } from "./base.js";

/**
 * Token Exceeds Limit Event Type
 */
export interface TokenLimitExceededEvent extends BaseEvent {
  type: "TOKEN_LIMIT_EXCEEDED";
  /** Number of Token currently in use */
  tokensUsed: number;
  /** Token Limit Threshold */
  tokenLimit: number;
}

/**
 * Token uses the warning event type
 */
export interface TokenUsageWarningEvent extends BaseEvent {
  type: "TOKEN_USAGE_WARNING";
  /** Number of Token currently in use */
  tokensUsed: number;
  /** Token Limit Threshold */
  tokenLimit: number;
  /** Percentage of use */
  usagePercentage: number;
}

/**
 * Error Event Type
 */
export interface ErrorEvent extends BaseEvent {
  type: "ERROR";
  /** Node ID (optional) */
  nodeId?: ID;
  /** error message */
  error: unknown;
  /** stack trace */
  stackTrace?: string;
}

/**
 * Variable Change Event Type
 */
export interface VariableChangedEvent extends BaseEvent {
  type: "VARIABLE_CHANGED";
  /** variable name */
  variableName: string;
  /** variable value */
  variableValue: unknown;
  /** variable scope */
  variableScope: string;
}

/**
 * LLM Streaming Abort Event Type
 */
export interface LLMStreamAbortedEvent extends BaseEvent {
  type: "LLM_STREAM_ABORTED";
  /** Node ID (optional) */
  nodeId?: ID;
  /** Reason for discontinuation */
  reason: string;
}

/**
 * LLM Streaming Error Event Type
 */
export interface LLMStreamErrorEvent extends BaseEvent {
  type: "LLM_STREAM_ERROR";
  /** Node ID (optional) */
  nodeId?: ID;
  /** error message */
  error: string;
}

/**
 * Context Compression Request Event Type
 */
export interface ContextCompressionRequestedEvent extends BaseEvent {
  type: "CONTEXT_COMPRESSION_REQUESTED";
  /** Number of Token currently in use */
  tokensUsed: number;
  /** Token Limit Threshold */
  tokenLimit: number;
  /** Dialogue statistics */
  stats?: unknown;
}

/**
 * Context Compression Completion Event Type
 */
export interface ContextCompressionCompletedEvent extends BaseEvent {
  type: "CONTEXT_COMPRESSION_COMPLETED";
  /** Summary of compression results */
  summary?: string;
  /** Number of Token after compression (optional) */
  tokensAfter?: number;
}

File: crates/wf-types/src/events/system_events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TokenLimitExceededEvent {
    pub base: super::BaseEvent,
    pub tokens_used: u32,
    pub token_limit: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TokenUsageWarningEvent {
    pub base: super::BaseEvent,
    pub tokens_used: u32,
    pub token_limit: u32,
    pub usage_percentage: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ErrorEvent {
    pub base: super::BaseEvent,
    pub node_id: Option<super::super::Id>,
    pub error: serde_json::Value,
    pub stack_trace: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VariableChangedEvent {
    pub base: super::BaseEvent,
    pub variable_name: String,
    pub variable_value: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LlmStreamAbortedEvent {
    pub base: super::BaseEvent,
    pub node_id: Option<super::super::Id>,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LlmStreamErrorEvent {
    pub base: super::BaseEvent,
    pub node_id: Option<super::super::Id>,
    pub error: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ContextCompressionRequestedEvent {
    pub base: super::BaseEvent,
    pub tokens_used: u32,
    pub token_limit: u32,
    pub stats: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ContextCompressionCompletedEvent {
    pub base: super::BaseEvent,
    pub summary: Option<String>,
    pub tokens_after: Option<u32>,
}
